File: 2024/senior/s4.py
from collections import defaultdict
from typing import Literal
from sys import setrecursionlimit
setrecursionlimit(300_000)

_, M = map(int, input().split())

graph = defaultdict(list)

for i in range(M):
    a, b = map(int, input().split())
    graph[a].append((b, i))
    graph[b].append((a, i))

seen = set()
paint = ['G'] * M

def dfs(node: int, color: Literal['R', 'B']) -> None:
    for neighbor, edge in graph[node]:
        if neighbor not in seen:
            seen.add(neighbor)
            paint[edge] = color
            dfs(neighbor, 'B' if color == 'R' else 'R')

for node in graph:
    if node not in seen:
        dfs(node, 'R')

print(''.join(paint))

File: 2024/senior/s4.java
import java.util.*;

public class s4 {
    static HashSet<Integer> seen = new HashSet<>();
    static HashMap<Integer, ArrayList<Integer[]>> graph = new HashMap<>();
    static char[] paint;

    static void dfs(int node, char color) {
        for (Integer[] neighbor : graph.get(node)) {
            int neighborNode = neighbor[0];

            if (!seen.contains(neighborNode)) {
                seen.add(neighborNode);
                paint[neighbor[1]] = color;
                dfs(neighborNode, color == 'R' ? 'B' : 'R');
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        sc.nextInt();

        int m = sc.nextInt();

        for (int i = 0; i < m; i++) {
            int a = sc.nextInt();
            int b = sc.nextInt();

            graph.putIfAbsent(a, new ArrayList<>());
            graph.putIfAbsent(b, new ArrayList<>());

            graph.get(a).add(new Integer[] {b, i});
            graph.get(b).add(new Integer[] {a, i});
        }

        sc.close();
        paint = new char[m];
        Arrays.fill(paint, 'G');

        for (int node : graph.keySet()) {
            if (!seen.contains(node)) {
                seen.add(node);
                dfs(node, 'R');
            }
        }

        System.out.println(new String(paint));
    }
}